### DISCUSSION TITLE
Airdrops in which the tokens are immediately transferred if the vesting period has ended

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/airdrops#33] [Ideas]
## Idea

As suggested by user, the idea is to provide an airdrop contract that transfers the assets directly to the user instead of creating a Sablier stream **when the vesting period has ended**.

Of course, this assumes that the airdrop is [ranged](https://github.com/sablier-labs/v2-periphery/discussions/279).

<img width="300" alt="SCR-20241223-qaqg" src="https://github.com/user-attachments/assets/8088c69c-d0b9-4c60-aac7-4bb74114fac3" />

## Feedback

cc @sablier-labs/engineers

Related:

- https://github.com/sablier-labs/v2-periphery/discussions/279
- https://github.com/sablier-labs/v2-core/discussions/912

**Comment by PaulRBerg** [comment]
I reckon that we could simply modify the existing Merkle contracts to enable this feature? Right, @sablier-labs/solidity?

Some changes would still be required in the UI to explain that the tokens will be received directly.

**Reply by smol-ninja** [reply]
We could do it through Merkle contracts, however, the calculations required to compute the `withdrawable` amount will be similar to the Lockup's implementation (think of Lockup Tranched based campaign). Therefore, I doubt if it will have much impact on the gas cost. We can also modify the `claim` function to simply call create followed by withdraw if the withdrawable amount is non-zero.

**Reply by andreivladbrg** [reply]
> I reckon that we could simply modify the existing Merkle contracts to enable this feature? 

yes, that's right.

---

> We could do it through Merkle contracts, however, the calculations required to compute the withdrawable amount will be similar to the Lockup's implementation (think of Lockup Tranched based campaign). Therefore, I doubt if it will have much impact on the gas cost. We can also modify the claim function to simply call create followed by withdraw if the withdrawable amount is non-zero

checking wether the withdrawble amount is non-zero and then `withdraw` sounds good to me

to make it more efficient, we could consider checking if the calculated end time in `_claim()` is equal to or less than `block.timestamp`. if it is, we could simply perform an ERC-20 transfer directly to the recipient without creating a stream and then withdrawing from it. the downside to this approach is that the recipient won’t receive an NFT if they claim after the end period .

**Reply by smol-ninja** [reply]
> we could consider checking if the calculated end time in `_claim()` is equal to or less than `block.timestamp`

Good idea.

**Reply by PaulRBerg** [reply]
There might have been a misunderstanding of what I suggested in the OP.

_vesting period has ended_ means exactly what @andreivladbrg said here: _checking if the calculated end time in `_claim()` is equal to or less than `block.timestamp`_

I never suggested performing partial withdraws when just a _portion_ has been vested.

So yes, the idea is to perform a basic ERC-20 transfer when `end time <= block timestamp`.

**Reply by smol-ninja** [reply]
My bad for misunderstanding the proposal's objective. Thank you for pointing it out. Looks good now.

**Comment by PaulRBerg** [comment]
Created issue for tracking: https://github.com/sablier-labs/airdrops/issues/37
